Fix removing remote shares when the remote server is unreachable

danxuliu opened the pull request:
Fixes #14797

To [delete a file through DAV](https://github.com/nextcloud/3rdparty/blob/c86fa3ed21786c646ef6c5e066a3a1b6f7505a71/sabre/dav/lib/DAV/Tree.php#L178-L179) the file info of the file is first got, but [if its storage is not available an exception is thrown](https://github.com/nextcloud/server/blob/65da8bd76e72dd94b882fa23e997617896f5eef2/apps/dav/lib/Connector/Sabre/Directory.php#L222-L224) and the deletion is aborted. If a file is a received federated share and the remote server is down the storage is not available, and therefore it is not possible to delete the file.

A received remote share can be deleted [using the RemoteController](https://github.com/nextcloud/server/blob/1a9330cd69631ef12b71239c4f98836e09e021ea/apps/files_sharing/lib/Controller/RemoteController.php#L178), even [if the remote instance is down](https://github.com/nextcloud/server/pull/5753). However, that OCS endpoint does not seem to be called from any UI element. Moreover, in order to use that endpoint the id of the remote share is needed, and in some cases it is not currently possible to [list the remote shares](https://github.com/nextcloud/server/blob/1a9330cd69631ef12b71239c4f98836e09e021ea/apps/files_sharing/lib/Controller/RemoteController.php#L141) when one of the remote instances is down (because in those cases [`getFileInfo`](https://github.com/nextcloud/server/blob/1a9330cd69631ef12b71239c4f98836e09e021ea/apps/files_sharing/lib/Controller/RemoteController.php#L119) throws an uncaught exception).

I have added integration tests for the issue, but I do not know how to fix it in the storage, so **I would need someone else to take over**. Sorry and thanks a lot! @rullzer @icewind1991 

**Pending:**
- [ ] Check [if kesselb's patch is the right approach](https://github.com/nextcloud/server/pull/25332#issuecomment-816176416) to fix the bug
  - [ ] If it is, check if the return code when deleting a file from an unreachable remote server should be 204 or 404
    - [ ] If it should be 204, adjust the fix
    - [ ] If it should be 404, adjust the test
  - [ ] If it is not, fix the bug with a different approach

## How to test

- In server A, open the Files app
- Share a file with user1 at server B
- In server B, login as user1
- Open the Files app and accept the remote share
- Kill server A
- In server B, try to delete the received file using the three dots menu

### Expected result

The file is deleted

### Actual result

The file can not be deleted

danxuliu committed:
Add integration tests for federated shares from unavailable servers

The federated server needs to be stopped during the tests, so it is now
stopped in the FederationContext for each scenario instead of just once
in the run.sh script.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for deleting federated shares

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

rullzer commented:
Master is Nextcloud 22 now.
If this should go into 21 it should be backported.

danxuliu commented:
/backport to stable21

danxuliu commented:
/backport to stable20

danxuliu commented:
/backport to stable19

danxuliu committed:
Add integration tests for listing federated shares

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
**Help needed**

I have rebased on latest master and added a few more test scenarios.

I tested the [change](https://github.com/nextcloud/server/issues/14797#issuecomment-568492497) suggested by @kesselb for [`getFileInfo`](https://github.com/nextcloud/server/blob/f98fed3f8a11bb71d4c10443b20a59c3864a3617/lib/private/Files/View.php#L1390) and it works (although `$data = null;` needs to be added in the `catch`, otherwise `$data` is undefined). 

However, although it is possible to delete the file using the Files app the last integration test does not pass even with that change. The problem is that it expects a 204 state when deleting the file, but it gets 404 instead. I guess that it makes sense that the file can not be found if the remote server is unreachable, but I do not know if that is just a side effect of the fix or if it is the right return code :shrug:

In any case, I do not know either if there is a better way to fix this or if that fix is the proper one. Hence the **help needed** :-)

PVince81 commented:
the tests reported as failing in the previous iteration were:
```
/drone/src/build/integration/federation_features/federated.feature:343
/drone/src/build/integration/federation_features/federated.feature:486
```

I've rebased this branch locally and ran all the new integration tests, they all passed for me.

Now pushed the rebased branch, let's see if they are now all green also on CI...

PVince81 commented:
so, the tests pass for me locally even on stable22 (with a local backport)

this is going to be tricky as it seems to be related to the env on which it is running
I'm guessing that the "server is down" detection might be triggering different behavior/exceptions depending on the setup which will make it harder to debug without relying on CI.

PVince81 commented:
the exception from CI:
```
    And remote server is stopped                                                                           # FederationContext::remoteServerIsStopped()
    When As an "user0"                                                                                     # FederationContext::asAn()
[Sat Sep 18 09:55:29 2021] {"Exception":"OCP\\Files\\StorageNotAvailableException","Message":"Sabre\\HTTP\\ClientException: Failed to connect to localhost port 8180: Connection refused","Code":1,"Trace":[{"file":"\/drone\/src\/lib\/private\/Files\/Storage\/DAV.php","line":289,"function":"convertException","class":"OC\\Files\\Storage\\DAV","type":"->","args":[{"__class__":"Sabre\\HTTP\\ClientException"},""]},{"file":"\/drone\/src\/lib\/private\/Files\/Storage\/DAV.php","line":785,"function":"propfind","class":"OC\\Files\\Storage\\DAV","type":"->","args":[""]},{"file":"\/drone\/src\/apps\/files_sharing\/lib\/External\/Storage.php","line":176,"function":"hasUpdated","class":"OC\\Files\\Storage\\DAV","type":"->","args":["",1631958927]},{"file":"\/drone\/src\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":381,"function":"hasUpdated","class":"OCA\\Files_Sharing\\External\\Storage","type":"->","args":["",1631958927]},{"file":"\/drone\/src\/lib\/private\/Files\/Storage\/Wrapper\/Availability.php","line":384,"function":"hasUpdated","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["",1631958927]},{"file":"\/drone\/src\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":381,"function":"hasUpdated","class":"OC\\Files\\Storage\\Wrapper\\Availability","type":"->","args":["",1631958927]},{"file":"\/drone\/src\/lib\/private\/Files\/Cache\/Watcher.php","line":127,"function":"hasUpdated","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["",1631958927]},{"file":"\/drone\/src\/lib\/private\/Files\/View.php","line":1351,"function":"needsUpdate","class":"OC\\Files\\Cache\\Watcher","type":"->","args":["",{"__class__":"OC\\Files\\Cache\\CacheEntry"}]},{"file":"\/drone\/src\/lib\/private\/Files\/View.php","line":1393,"function":"getCacheEntry","class":"OC\\Files\\View","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Storage","cache":null,"scanner":null,"watcher":null,"propagator":null,"updater":null},"","\/remote-share.txt"]},{"file":"\/drone\/src\/apps\/files_sharing\/lib\/Controller\/RemoteController.php","line":118,"function":"getFileInfo","class":"OC\\Files\\View","type":"->","args":["\/user0\/files\/remote-share.txt"]},{"function":"extendShareInfo","class":"OCA\\Files_Sharing\\Controller\\RemoteController","type":"::","args":[{"id":"19","share_type":"0","parent":"-1","remote":"http:\/\/localhost:8180\/","remote_id":"19","0":"And 6 more entries, set log level to debug to see all entries"}]},{"file":"\/drone\/src\/apps\/files_sharing\/lib\/Controller\/RemoteController.php","line":142,"function":"array_map","args":["self::extendShareInfo",[{"id":"19","share_type":"0","parent":"-1","remote":"http:\/\/localhost:8180\/","remote_id":"19","0":"And 6 more entries, set log level to debug to see all entries"}]]},{"file":"\/drone\/src\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":217,"function":"getShares","class":"OCA\\Files_Sharing\\Controller\\RemoteController","type":"->","args":[]},{"file":"\/drone\/src\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":126,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Files_Sharing\\Controller\\RemoteController"},"getShares"]},{"file":"\/drone\/src\/lib\/private\/AppFramework\/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Files_Sharing\\Controller\\RemoteController"},"getShares"]},{"file":"\/drone\/src\/lib\/private\/Route\/Router.php","line":301,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\Files_Sharing\\Controller\\RemoteController","getShares",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"ocs.files_sharing.Remote.getShares"}]},{"file":"\/drone\/src\/ocs\/v1.php","line":62,"function":"match","class":"OC\\Route\\Router","type":"->","args":["\/ocsapp\/apps\/files_sharing\/api\/v1\/remote_shares"]}],"File":"\/drone\/src\/lib\/private\/Files\/Storage\/DAV.php","Line":864,"Hint":"Storage is temporarily not available","CustomMessage":"--"}
[Sat Sep 18 09:55:29 2021] 127.0.0.1:48998 [200]: /ocs/v1.php/apps/files_sharing/api/v1/remote_shares
```

we see this happening as soon as user0 logs in.

Seeing the exception reminded me of the "availability" wrapper which buffers availability checks. Basically, if it sees that a remote was not available it stores that information for about 10 minutes. So the next checks will use the cached result for that time.

And maybe depending on the environment and test order and speed, the availability caching might behave differently, which can explain why a failure can sometimes happen in different parts of the code.

danxuliu commented:
The integration tests always fail for me when locally run, provided [I am running them through Docker](https://github.com/nextcloud/server/blob/3cda67c8b13e4841669e50f0cc019f3cd0afdd71/build/integration/run-docker.sh).

You can do that with `./run-docker.sh federation_features/federated.feature` from `build/integration/` (but please check the script to know about the requirements). The tests also fail if a single test is run with `./run-docker.sh federation_features/federated.feature:343` or `./run-docker.sh federation_features/federated.feature:486`.

PVince81 commented:
@danxuliu good news: you were right that running with docker can recreate the issues, I was able to do so locally:
```
    /nextcloud/build/integration/federation_features/federated.feature:343
    /nextcloud/build/integration/federation_features/federated.feature:486
```

PVince81 committed:
Return false in hasUpdated when storage is not available

Technically, saying that a storage has no updates when it's not
available is correct.

This makes it possible to retrieve the cache entry for the mount point
and also to list and remove unavailable federated shares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Update psalm baseline

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

skjnldsv commented:
> got confirmation in chat that this is fine

Let's review and merge then! :rocket:

PVince81 commented:
rebased to make this more up to date

PVince81 commented:
Hmmm, I've compared the behavior on master with this PR and now when a storage is not available, it is still possible to explore its folder tree. But any write operation will fail with StorageNotAvailableException.
Which means that the "gatekeeping" was mostly done by the `hasUpdated` method

This feels like a big improvement because it makes it possible to still view the cache tree instead of feeling like being locked out.

PVince81 committed:
Update lib/private/Files/Storage/Wrapper/Availability.php add void

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

Co-authored-by: Carl Schwan <carl@carlschwan.eu>

acsfer commented:
/backport to stable22
